### PR TITLE
Add support for custom hyper-dash catcher colouring

### LINKED ISSUE #8085: LegacySkinDecoder does not support [CatchTheBeat] sections
For reference: https://osu.ppy.sh/help/wiki/Skinning/skin.ini#[catchthebeat]

### PR DESCRIPTION
**Pull request by frenzibyte**
- [x] Depends on #8446 (requires merge conflict resolution)

Closes #8085

### CODE REVIEW

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs]: Same as the other PR, these step names are not really legible.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs]: Not sure I like these being loose and out in the open in the catcher like this. Might make sense to encapsulate in some sort of private subcontainer class to reduce noise in the catcher class.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs]: Comment basically restates the code.

File: osu.Game.Rulesets.Catch/UI/CatcherArea.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/CatcherArea.cs]: If it's unsafe to call in the object initialiser it shouldn't be a property at all...? I'd say this needs to be rectified, as it looks like a foot-gun.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs]: Did you check what are the implications of no longer checking hyper-dash state on the previous frame and skipping fades if there has been no change, as was done previously? The previous way would avoid redundant colour fades/invalidations if consecutive `updateCatcherColour()` calls were made that would not change value of `HyperDashing`.

**Review by bdach — DISMISSED** [review]

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:43]: just to be clear, do you mean to put these 3 trails containers inside a private container with public accessibility to all containers inside and access them on all usages?

**Comment by bdach** [comment]
Going to hold off on reviewing this until the prereq is dealt with as due to the merging in-between it's difficult to see what the actual change of this PR is and give feedback on it.

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:160]: these assert texts are too long. should be a few words

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:144]: these are too long. remove unnecessary words (the class already asserts most of this, and "custom" is redundant since you already say "fallback")

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:292]: why nt switch

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:275]: bool bool bool is so hard to read. why not just make the custom color4 properties nullable and settable and set them in the tests? at least the tests would be legible then.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
review applies to previous pr i guess. not going to repost there

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:160]: So the assert names used to be like: `end-glow sprite colour default-hyperdash`, but I was change-requested to make it more described as the assert name are mostly for describing the assertion on CI in case it fails, although I agree that it is overly-described and it is my mistake not pointing that out in the request conversation.

Comment (peppy) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:160]: "colour is correct" is fine. if you really need to explain more that isn't obvious from the test method's name, add an inline comment.

**Comment by bdach** [comment]
Tried to see if I can merge master back in here but there's been so many changes it's nigh unmergeable. It would likely be easier to just rewrite this PR off master at this point.

**Comment by iiSaLMaN** [comment]
> Tried to see if I can merge master back in here but there's been so many changes it's nigh unmergeable. It would likely be easier to just rewrite this PR off master at this point.

It was actually just a conflicting of the new written `TestSceneHyperDashColouring` so I fixed it, should be good now.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:29]: this doesn't read well. also please document the above colour if you're going to do so here.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:51]: I think this is fine.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:162]: just pass the target into the `ctor` and remove all of this.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:172]:
```suggestion
                dashTrails ??= new Container<CatcherTrailSprite> { RelativeSizeAxes = Axes.Both },
```

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:174]: initialise all of this in the ctor or bdl

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:136]: this is completely wrong

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:32]: Not sure if this is me being dense again but I still kind of don't get what this particular xmldoc is trying to say... In its current state it could be interpreted in a way that makes it misleading (as in, that this colour applies to the whole catcher, which is not *technically* the case - it applies only in hyper-dash state, even if we use `Colour` for that due to additive blending).

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:136]: I'm not quite sure how to test this correctly without directly exposing the containers...

I would say applying what @bdach suggested regarding grouping all trail containers in one `CatcherTrailsContainer` may help this by doing:
```csharp
catcherArea.OfType<CatcherTrailsContainer>().Single().HyperDashTrails == expectedColour
```

Comment (peppy) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:136]: Try it and see how it feels. Could work.

Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs]: After creating a public `CatcherTrailsDisplay` grouping all trails, I've decided to also move the trail-related logic into that component now that there's a chance to, for simplifying `Catcher`.

File: osu.Game.Rulesets.Catch.Tests/TestSceneCatcher.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneCatcher.cs:27]: as mentioned in discord, this container should be displayed in the test scene. you've regressed the test as it stands.

Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch.Tests/TestSceneCatcher.cs:27]: This test used to provide `null` for `additiveTarget` so I thought there won't be a problem with providing a dummy `Container`? 

Moreover, providing a real `additiveTarget` here sounds exactly like what `CatcherArea` provides, which has an own test scene (`TestSceneCatcherArea`), why not just remove this test scene if it's worthless.

Comment (peppy) [reply osu.Game.Rulesets.Catch.Tests/TestSceneCatcher.cs:27]: i guess it's fine for this case, as input is also not handled here.

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:129]: can this be replaced with an until step?

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:133]: i don't understand this ternary conditional.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:34]: I still think this xmldoc is misleading (if not unreadable - I still don't quite understand what it's trying to say), and more importantly, it directly contradicts the xmldoc of `DEFAULT_HYPER_DASH_COLOUR` immediately preceding it. I've mentioned this once and I'm pretty sure I unresolved the relevant conversation, but the conversation is resolved again and I see no actual changes to the content of the xmldoc, save for insertion of one new line.

Please don't resolve conversations without addressing them in some shape or form.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:295]: I find this behaviour of silently overriding the colour from `Red` to `OrangeRed` really confusing. It complicates everything (including test code) and leads to unexpected behaviour (the user setting a red colour is now indistinguishable from the default - in both cases this colour will get silently converted to `OrangeRed`). I propose we just get rid of it for simplicity.

Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:295]: I agree, I've tried comparing them with each other, the `Red` colour made the sprite more red-er than `OrangeRed`, but since the catcher is small I don't think it'll be that notable.

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:138]: this still does not make any sense to me, sorry. at this point you should be splitting out tests rather than doing this weird logic.

- what is "another colour" and why?
- why not just set it correctly to `expectedCatcherColour`?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:295]: i'm fine with just removing it if you can't figure out how to write tests. your call.

Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:295]: The only 2 sane ways I can provide to implement this would be:
 1. Do some type of calculation to get a colour that can blend with the trails similar to how `Color4.Red` blends with `Color4.OrangeRed(alpha: 0.2f)`, sounds nearly impossible or just way too complicated to even be done correctly, *not even sure to call it a sane way*.
2. Rename `hyperDashColour` to `customHyperDashColour` and let it fallback to `DEFAULT_CATCHER_HYPER_DASH_COLOUR` on null here, that's a bit more readable and will solve the issue @bdach stated above, but the effect will only come here. as for the tests, I'll replay in the below conversation about it.

If nothing here sounds sane to you then I honestly don't know, may be that I can't really be more helpful in this specific case or that this is just way too complicated to be simplified.

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:138]: What do you mean by "splitting out tests", might be what I'm gonna suggest here:

I guess the best this can become would be by splitting `expectedCatcherColour` to `expectedCatcherColour, expectedTrailsColour`, and fill in accordingly.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (iiSaLMaN) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:34]:
```suggestion
        /// <summary>
        /// The default color used for the direct <see cref="Drawable.Colour"/> of the <see cref="Catcher"/> while in hyper-dashing state.
        /// </summary>
        /// <remarks>
        /// This colour is used specially in here on the default skin only.
        /// </remarks>
```

What do you think?

File: osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs
Comment (bdach) [reply osu.Game.Rulesets.Catch.Tests/TestSceneHyperDashColouring.cs:138]: Basically if your test assertions have conditionals in them it means it's time to split the test into two, each of which tests each branch of the conditional assertion - so you'd test the `DEFAULT_HYPER_DASH_COLOUR` case separately and the case of a custom non-red colour separately.

That said I'd still be in favour of just getting rid of that special case.

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (bdach) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs:34]: No, I still don't understand what that xmldoc is trying to say. What is a "*direct colour*"? You're still not describing what the colours are actually *colouring*.

If I had to have a go at documenting these constants somewhat accurately (from what I can tell from a quick look), I'd probably write something along the lines of

```csharp
/// <summary>
/// The default colour used to tint hyper-dash fruit, the catcher's trail and its after-image during a hyper-dash.
/// </summary>
public static readonly Color4 DEFAULT_HYPER_DASH_COLOUR = Color4.Red;

/// <summary>
/// The default colour used to tint the moving catcher sprite, along with the fruits on the catcher plate, during a hyper-dash.
/// </summary>
public static readonly Color4 DEFAULT_CATCHER_HYPER_DASH_COLOUR = Color4.OrangeRed;
```

which to me shows quite well how arbitrary and weird this distinction has become (unless I'm wrong about this description, but then it shows that this colour rigamarole is way too complicated).

**Comment by bdach** [comment]
I went ahead and removed the `DEFAULT_CATCHER_HYPER_DASH_COLOUR` thing entirely myself. It feels like an okay thing to do as it simplifies matters significantly while having barely noticeable (at least to me) visual implications.

The rest of this seems relatively sane as far as I can tell.

File: osu.Game.Rulesets.Catch/UI/CatcherTrailDisplay.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/CatcherTrailDisplay.cs:106]: I find it super weird that `Catcher` is responsible for starting the trail and this new class is responsible for stopping it...

File: osu.Game.Rulesets.Catch/UI/Catcher.cs
Comment (peppy) [inline osu.Game.Rulesets.Catch/UI/Catcher.cs:77]: this is horrible logic. you can't arbitrarily or in a value in a set property. the chances of this breaking are 99.9% (it is not going to be applied unless `Dashing` is set, but there's a precondition equality check potentially causing the set to be a no-op).

Comment (bdach) [reply osu.Game.Rulesets.Catch/UI/Catcher.cs]: Would you mind having a look at the commits I just pushed in here and seeing if it improves things? I'm not 100% sure it does but after looking at this code for the better part of an hour I decided to avoid any bigger changes for now.

**Review by peppy — APPROVED** [review]
seems sane enough now

**Pull request merged**
